src/values/ObjectValue.js

```diff
@@ -481,6 +491,21 @@ export default class ObjectValue extends ConcreteValue {
     return keyArray;
   }
 
+  getSnapshot(): AbstractObjectValue {
+    if (this.temporalAlias !== undefined) {
+      this.properties = new Map();
+      return this.temporalAlias;
+    }
+    let template = new ObjectValue(this.$Realm);
+    Object.assign(template, this);
+    delete template.hashValue;
+    this.properties = new Map();
```

Comment:
I don’t know if getting a snapshot should imply losing information about all properties.

I think of the typical use case for a snapshot as calling a pure function with a mutable object.

```js
var obj = {foo: “x”};
var x = pure(obj);
obj.foo = “y”;
var y = pure(obj);
```

In these cases we don’t want to forget about the properties. It seems to me that forgetting is a special case.

It also seems odd to only forget properties and not any other internal values, prototypes etc.

Reply:
Btw, Object.assign works on symbols too. So you'd need to do this for the symbols map too.

Reply:
There are scenarios where we want a snapshot without forgetting about the properties. This should be handled by the caller, for the sake of efficiency, but also for flexibility. I don't understand your point about calls to pure functions. Currently, any parameter passed to an abstract function is havoced and that seems appropriate. It seems to me that a havoced object should not need its properties. Let's get back to this in a later PR.

I think symbols will work just fine the way things are right now, but I'll add some test cases just to make sure.

Reply:
We've talked about, in the future, handling abstract functions that are marked as pure (not in a pure scope but just calling a pure function anytime). Those should not need to havoc.

To implement that, we'd need some form of snapshots.

Reply:
I think that we can handle that scenario without rewriting getSnapshot. I guess we'll see, but for now I'd like to go ahead with the current formulation.

Reply:
Re: symbols. Prepack has currently very poor support for symbols. We also require partial targets and sources to be simple, which imply no symbols. We do not currently fail if you create a symbol property on an intrinsic or temporal object. I need to get back to this, but it seems inappropriate to fix this in the current pull request.